### PR TITLE
Add multi-session test

### LINKED ISSUE #1105: [BUG] Test: ManagedChannel not shutdown properly in CI
**Describe the bug**

CI fails with `ManagedChannel not shutdown properly`, follows by `table of database not exists`. The two exceptions may be linked.
```
[2019-09-16T19:14:50.742Z] RowIDAllocatorSuite:

[2019-09-16T19:14:50.742Z] 19/09/17 03:14:50 INFO SharedSQLContext: load data is mode: auto

[2019-09-16T19:14:50.742Z] 19/09/17 03:14:50 ERROR ManagedChannelOrphanWrapper: *~*~*~ Channel ManagedChannelImpl{logId=401, target=127.0.0.1:20160} was not shutdown properly!!! ~*~*~*

[2019-09-16T19:14:50.742Z]     Make sure to call shutdown()/shutdownNow() and wait until awaitTermination() returns true.

[2019-09-16T19:14:50.742Z] java.lang.RuntimeException: ManagedChannel allocation site

[2019-09-16T19:14:50.742Z] 	at io.grpc.internal.ManagedChannelOrphanWrapper$ManagedChannelReference.<init>(ManagedChannelOrphanWrapper.java:103)

[2019-09-16T19:14:50.742Z] 	at io.grpc.internal.ManagedChannelOrphanWrapper.<init>(ManagedChannelOrphanWrapper.java:53)

[2019-09-16T19:14:50.742Z] 	at io.grpc.internal.ManagedChannelOrphanWrapper.<init>(ManagedChannelOrphanWrapper.java:44)

[2019-09-16T19:14:50.742Z] 	at io.grpc.internal.AbstractManagedChannelImplBuilder.build(AbstractManagedChannelImplBuilder.java:419)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.util.ChannelFactory.lambda$getChannel$0(ChannelFactory.java:49)

[2019-09-16T19:14:50.742Z] 	at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.util.ChannelFactory.getChannel(ChannelFactory.java:34)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.region.RegionStoreClient$RegionStoreClientBuilder.build(RegionStoreClient.java:581)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.region.RegionStoreClient$RegionStoreClientBuilder.build(RegionStoreClient.java:592)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.operation.iterator.ConcreteScanIterator.loadCurrentRegionToCache(ConcreteScanIterator.java:75)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.operation.iterator.ScanIterator.cacheLoadFails(ScanIterator.java:78)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.operation.iterator.ConcreteScanIterator.hasNext(ConcreteScanIterator.java:103)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.codec.MetaCodec.hashGetFields(MetaCodec.java:103)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.catalog.CatalogTransaction.getTables(CatalogTransaction.java:73)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.catalog.Catalog$CatalogCache.loadTables(Catalog.java:105)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.catalog.Catalog$CatalogCache.lambda$loadDatabases$1(Catalog.java:124)

[2019-09-16T19:14:50.742Z] 	at java.lang.Iterable.forEach(Iterable.java:75)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.catalog.Catalog$CatalogCache.loadDatabases(Catalog.java:119)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.catalog.Catalog$CatalogCache.<init>(Catalog.java:56)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.catalog.Catalog$CatalogCache.<init>(Catalog.java:50)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.catalog.Catalog.reloadCache(Catalog.java:163)

[2019-09-16T19:14:50.742Z] 	at com.pingcap.tikv.catalog.Catalog.lambda$new$0(Catalog.java:148)

[2019-09-16T19:14:50.742Z] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)

[2019-09-16T19:14:50.742Z] 	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)

[2019-09-16T19:14:50.742Z] 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)

[2019-09-16T19:14:50.742Z] 	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)

[2019-09-16T19:14:50.742Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)

[2019-09-16T19:14:50.742Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)

[2019-09-16T19:14:50.742Z] 	at java.lang.Thread.run(Thread.java:748)

[2019-09-16T19:14:50.999Z] - test unsigned allocator

[2019-09-16T19:14:51.258Z] - test signed allocator *** FAILED ***

[2019-09-16T19:14:51.258Z]   java.lang.IllegalArgumentException: table or database is not existed

[2019-09-16T19:14:51.258Z]   at com.pingcap.tikv.allocator.RowIDAllocator.getAutoTableId(RowIDAllocator.java:182)

[2019-09-16T19:14:51.258Z]   at com.pingcap.tikv.allocator.RowIDAllocator.initSigned(RowIDAllocator.java:188)

[2019-09-16T19:14:51.258Z]   at com.pingcap.tikv.allocator.RowIDAllocator.create(RowIDAllocator.java:54)

[2019-09-16T19:14:51.258Z]   at com.pingcap.tispark.datasource.RowIDAllocatorSuite$$anonfun$2.apply(RowIDAllocatorSuite.scala:39)

[2019-09-16T19:14:51.258Z]   at com.pingcap.tispark.datasource.RowIDAllocatorSuite$$anonfun$2.apply(RowIDAllocatorSuite.scala:25)

[2019-09-16T19:14:51.258Z]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)

[2019-09-16T19:14:51.258Z]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)

[2019-09-16T19:14:51.258Z]   at org.scalatest.Transformer.apply(Transformer.scala:22)

[2019-09-16T19:14:51.258Z]   at org.scalatest.Transformer.apply(Transformer.scala:20)

[2019-09-16T19:14:51.258Z]   at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:186)

[2019-09-16T19:14:51.258Z]   ...
```

**Spark and TiSpark version info**

Spark 2.3, TiSpark Master

**Additional context**
TiDB cluster version v2.1.17

### LINKED ISSUE #1215: [BUG] Table not exist bug
https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tispark_ghpr_integration_test/detail/tispark_ghpr_integration_test/1125/pipeline

set 
```
region-max-size = "2MB"
region-split-size = "1MB"
```
and run unit test, got the following error

```
[2019-11-14T03:17:33.464Z] 19/11/14 11:17:33 INFO SharedSQLContext: load data is mode: auto

[2019-11-14T03:18:41.257Z] - test unique indices cases *** FAILED ***

[2019-11-14T03:18:41.257Z]   com.pingcap.tikv.exception.TiClientInternalException: Table not exist TiTableReference(tidb_batch_write_test_index,batch_write_insertion_one_unique_index56,9223372036854775807)

[2019-11-14T03:18:41.257Z]   at com.pingcap.tispark.TiDBRelation$$anonfun$1.apply(TiDBRelation.scala:41)

[2019-11-14T03:18:41.257Z]   at com.pingcap.tispark.TiDBRelation$$anonfun$1.apply(TiDBRelation.scala:41)

[2019-11-14T03:18:41.257Z]   at scala.Option.getOrElse(Option.scala:121)

[2019-11-14T03:18:41.257Z]   at com.pingcap.tispark.TiDBRelation.<init>(TiDBRelation.scala:41)

[2019-11-14T03:18:41.257Z]   at com.pingcap.tispark.TiDBDataSource.createRelation(TiDBDataSource.scala:45)

[2019-11-14T03:18:41.257Z]   at com.pingcap.tispark.TiDBDataSource.createRelation(TiDBDataSource.scala:69)

[2019-11-14T03:18:41.258Z]   at org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand.run(SaveIntoDataSourceCommand.scala:45)

[2019-11-14T03:18:41.258Z]   at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult$lzycompute(commands.scala:70)

[2019-11-14T03:18:41.258Z]   at org.apache.spark.sql.execution.command.ExecutedCommandExec.sideEffectResult(commands.scala:68)

[2019-11-14T03:18:41.258Z]   at org.apache.spark.sql.execution.command.ExecutedCommandExec.doExecute(commands.scala:86)
```

### PR DESCRIPTION
**Pull request by birdstorm**
### What problem does this PR solve? 
Add multi-session tests as supplement to #1248.
The changes found in test framework will also close #1215, and close #1105 

### Check List 

Tests 

 - Unit test

Code changes

 - Has interface methods change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to be included in the release note

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/sql/test/SharedSQLContext.scala [hunk 1]
```diff
@@ -200,38 +167,21 @@ object SharedSQLContext extends Logging {
       }
 
       if (_ti != null) {
-        _ti.sparkSession.sessionState.catalog.reset()
-        _ti.meta.close()
-        _ti.sparkSession.close()
-        _ti.tiSession.close()
+        _ti.meta.reloadMeta()
```
Comment (marsishandsome) [inline core/src/test/scala/org/apache/spark/sql/test/SharedSQLContext.scala]: delete this line?

File: core/src/test/scala/org/apache/spark/SharedSparkContext.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+trait SharedSparkContext extends BeforeAndAfterAll with BeforeAndAfterEach { self: Suite =>
+
+  def sc: SparkContext = _sc
+
+  protected var conf: SparkConf = new SparkConf(false)
+
+  protected def initializeContext(): Unit = {
+    if (null == _sc) {
+      _sc = new SparkContext("local[4]", "test", conf)
```
Comment (marsishandsome) [inline core/src/test/scala/org/apache/spark/SharedSparkContext.scala]: why `local[4]`?

Comment (birdstorm) [reply core/src/test/scala/org/apache/spark/SharedSparkContext.scala]: same as Spark Core test

File: core/src/test/scala/org/apache/spark/sql/test/SharedSQLContext.scala (hunk 1 above)
Comment (birdstorm) [reply core/src/test/scala/org/apache/spark/sql/test/SharedSQLContext.scala]: prepare to delete it later.

File: core/src/test/scala/org/apache/spark/sql/test/TestSparkSession.scala [hunk 3]
```diff
@@ -15,19 +15,20 @@
...
+private[spark] class TestSparkSession(sparkContext: SparkContext, isHiveEnabled: Boolean) {
   self =>
   private val builder = SparkSession
     .builder()
-    .master("local[*]")
+    .sparkContext(sparkContext)
+    .master("local[4]")
     .appName("tispark-integration-test")
-    .config(sparkConf.set("spark.sql.testkey", "true"))
+    .config("spark.sql.test.key", "true")
```
Comment (marsishandsome) [inline core/src/test/scala/org/apache/spark/sql/test/TestSparkSession.scala]: why change this name?

File: core/src/test/scala/org/apache/spark/SharedSparkContext.scala (hunk 2 above)
Comment (marsishandsome) [reply core/src/test/scala/org/apache/spark/SharedSparkContext.scala]: cannot find `local[4]` in current code base

File: core/src/test/scala/org/apache/spark/sql/test/SharedSQLContext.scala (hunk 1 above)
Comment (marsishandsome) [reply core/src/test/scala/org/apache/spark/sql/test/SharedSQLContext.scala]: why not in this PR?

Comment (birdstorm) [reply core/src/test/scala/org/apache/spark/sql/test/SharedSQLContext.scala]: i will delete it if i can pass the CI

File: core/src/test/scala/org/apache/spark/SharedSparkContext.scala (hunk 2 above)
Comment (birdstorm) [reply core/src/test/scala/org/apache/spark/SharedSparkContext.scala]: https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/SharedSparkContext.scala#L42

File: core/src/test/scala/org/apache/spark/sql/test/TestSparkSession.scala (hunk 3 above)
Comment (birdstorm) [reply core/src/test/scala/org/apache/spark/sql/test/TestSparkSession.scala]: because sparkConf.set is ambiguous

Comment (marsishandsome) [reply core/src/test/scala/org/apache/spark/sql/test/TestSparkSession.scala]: i mean why change `spark.sql.testkey` to `spark.sql.test.key` ?

Comment (birdstorm) [reply core/src/test/scala/org/apache/spark/sql/test/TestSparkSession.scala]: Good point, I think `spark.sql.test.key` is better

**Comment by birdstorm** [comment]
/run-all-tests tidb=release-3.0 tikv=release-3.0 pd=release-3.0 mode=full

File: core/src/test/scala/org/apache/spark/sql/MultipleSparkSessionTestSuite.scala [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+class MultipleSparkSessionTestSuite extends BaseTiSparkTest {
+  test("Multiple Spark Session register udf") {
+    val sparkSession1 = spark
+    assert(sparkSession1.sql("select ti_version()").count() === 1)
+    SparkSession.clearActiveSession()
```
Comment (marsishandsome) [inline core/src/test/scala/org/apache/spark/sql/MultipleSparkSessionTestSuite.scala]: can we delete this line ?

File: core/src/test/scala/org/apache/spark/sql/MultipleSparkSessionTestSuite.scala [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    SparkSession.clearDefaultSession()
```
Comment (marsishandsome) [inline core/src/test/scala/org/apache/spark/sql/MultipleSparkSessionTestSuite.scala]: can we delete this line ?

File: core/src/test/scala/org/apache/spark/sql/MultipleSparkSessionTestSuite.scala (hunk 4 above)
Comment (birdstorm) [reply core/src/test/scala/org/apache/spark/sql/MultipleSparkSessionTestSuite.scala]: No, this is essential to test multi spark session.

File: core/src/test/scala/org/apache/spark/sql/MultipleSparkSessionTestSuite.scala (hunk 5 above)
Comment (birdstorm) [reply core/src/test/scala/org/apache/spark/sql/MultipleSparkSessionTestSuite.scala]: maybe this line can be deleted, but I add it for better understanding when something goes wrong.

**Comment by birdstorm** [comment]
/run-all-tests tidb=release-3.0 tikv=release-3.0 pd=release-3.0

File: core/src/test/scala/org/apache/spark/sql/MultipleSparkSessionTestSuite.scala [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    assert(sparkSession2.sql("select ti_version()").count() === 1)
+  }
+
+  test("Test multiple Spark Session with different catalog") {
+    val sparkSession1 = spark
+    val sparkSession2 = sparkSession1.newSession()
+    sparkSession1.sql(s"use ${dbPrefix}tispark_test")
+    sparkSession2.sql(s"use ${dbPrefix}prefix_index")
+    sparkSession1.sql("show tables").show()
+    sparkSession2.sql("show tables").show()
```
Comment (marsishandsome) [inline core/src/test/scala/org/apache/spark/sql/MultipleSparkSessionTestSuite.scala]: add some asserts here?

Comment (birdstorm) [reply core/src/test/scala/org/apache/spark/sql/MultipleSparkSessionTestSuite.scala]: done

**Pull request merged**